### ISSUE TITLE
Useless assertion in twisted.web.microdom.Node.replaceChild

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3421]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3421|
|Type|defect|
|Created|2008-09-08 13:39:26Z|
|Branch|https://github.com/twisted/twisted/tree/microdom-assert-3421|

The assertion is always positive.  Plus asserts are almost always pointless.

Attachments:

* [microdom.assert.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/872/872f4393d69bcbe4912f49ece93faa991d404b91/65e703a811c7d44addce05f38275b4ef2208e96c.diff) (668 bytes) - added by jparise on 2008-11-18 07:16:58Z - Corrects assert syntax
* [microdom.assert2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/872/872f4393d69bcbe4912f49ece93faa991d404b91/9a279cee2e07d17eb20b5afe65d9550ce18cfade.diff) (4492 bytes) - added by jparise on 2008-11-24 07:19:12Z - Replaces asserts with exceptions

<details><summary>Searchable metadata</summary>

```
trac-id__3421 3421
type__defect defect
reporter__exarkun exarkun
priority__low low
milestone__Python_2_6 Python-2.6
branch__branches_microdom_assert_3421 branches/microdom-assert-3421
branch_author__exarkun__jparise exarkun, jparise
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1220881166000000 1220881166000000
changetime__1228229244000000 1228229244000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="jparise's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|jparise commented|
|-|-|

The attached patch corrects the syntax of the offending `assert` statement.  This `assert` now functions as intended and no longer generates warnings under Python 2.6+.  There should be no other behavioral differences.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun set owner to jparise|
|-|-|

Asserts are silly.  Can you instead change it to an explicit check which raises some documented interface and add a unit test for the behavior?

**Comment by twisted-trac** [comment]
|<img alt="jparise's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|jparise commented|
|-|-|

I've attached a new patch that replaces a handful of assertions with exceptions.  The unit test has been updated, and the exceptions are documented in the methods' docstrings.

There are still some remaining assertions in microdom.py that could also be replaced with exceptions.  Those cases are more "interesting", however, as they would benefit most from custom exception classes given their microdom-ish-specific nature.  If there's a desire to convert those as well, it might be best to track that work under a separate ticket.

**Comment by twisted-trac** [comment]
|<img alt="jparise's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|jparise commented|
|-|-|

Adding the Python-2.6 milestone because the current trunk code produces a warning under Python 2.6.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [25490]) Branching to 'microdom-assert-3421'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [25491]) apply microdom.assert2.diff

refs [#3421](https://github.com/twisted/twisted/issues/3421)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [25493]) split replaceChild ValueError test into separate method

refs [#3421](https://github.com/twisted/twisted/issues/3421)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [25494]) remove useless dom parent fakes

refs [#3421](https://github.com/twisted/twisted/issues/3421)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [25495]) raise ValueError from Document.appendChild if there are too many children

refs [#3421](https://github.com/twisted/twisted/issues/3421)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [25496]) some docstrings

refs [#3421](https://github.com/twisted/twisted/issues/3421)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

thanks jparise.

I've made a few more changes in the branch.  Someone else should take a look at this now.

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_14"></a>|@mwhudson commented|
|-|-|

The exception in the branch make more sense than the assertions in trunk.  They're still not very good, as they don't really provide much help in fixing your problem (like, what type was it if it wasn't a Node?  For replaceChild, which parameter was it?).

OTOH, microdom is on it's way out anyway (right?), so I don't see much point in expending effort in fixing the above.

Land the branch, IOW.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set status to `closed`|
|-|-|

(In [25554]) Merge microdom-assert-3421

Author: jparise, exarkun
Reviewer: exarkun, mwhudson
Fixes: [#3421](https://github.com/twisted/twisted/issues/3421)

Remove most asserts from microdom and replace them with checks which raise some
documented exception.  In particular, remove one assertion which could never
fail and which produced a warning on Python 2.6.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
